torch/utils/_mode_utils.py

```diff
@@ -0,0 +1,62 @@
+import torch
+from typing import Iterator
+from torch._C import (
+    _get_torch_function_mode, _set_torch_function_mode, _get_python_mode, _set_python_mode)
+
+# a class for the helper function to package all the info about the Mode class
+# so the helper function can access it without needing a circular import
+class _ModeInfo:
+    def __init__(self, mode_type: str, mode_class: type, base_mode_class: type, mode_class_name: str):
+        if mode_type not in ['torch_function', 'python']:
+            raise RuntimeError(f"only support torch_function or python modes, got mode_type of {mode_type}")
+        self.is_torch_function_mode = mode_type == 'torch_function'
+        self.mode_type = mode_type
+        self.mode_class = mode_class
+        self.base_mode_class = base_mode_class
+        self.mode_class_name = mode_class_name
+
+def _enable_mode(mode, mode_info: _ModeInfo, *, replace=None, ignore_preexisting=False) -> Iterator[None]:
+    if mode_info.is_torch_function_mode:
+        allowed_types = (
+            isinstance(mode, mode_info.mode_class) or
+            (isinstance(mode, type) and not issubclass(mode, mode_info.mode_class)))
+    else:
+        allowed_types = isinstance(mode, type) and issubclass(mode, (torch.Tensor,))
+    if not (mode is None or allowed_types):
+        if mode_info.is_torch_function_mode:
+            allowed_options = "TorchFunctionMode, Tensor-like class, or None"
+        else:
+            allowed_options = "Tensor-like class or None"
+        raise ValueError(f'expected to get {allowed_options} as argument, got {type(mode)} instead')
+    old = _get_torch_function_mode() if mode_info.is_torch_function_mode else _get_python_mode()
+    if old is mode:
```

Comment:
I don't know if this is overkill (trying to merge the implementations of enable_torch_function_mode and enable_torch_dispatch_mode). We're not sure what the long-term interaction of both of these things together will look like.

If you do want to go down this route, having all of these if-then-else-blah is not the cleanest thing in the world. A better idiom might be to have:
- ModeInfo class that has more extension points (e.g., it can have a "supports_input(input) -> bool" method
- TorchFunctionModeInfo class that subclasses ModeInfo with all of the overrides
- TorchDispatchModeInfo class that subclasses ModeInfo
- _enable_mode implementation accepts a ModeInfo class and only queries attributes or calls methods on ModeInfo. It shouldn't need to if/then/else switching on whether the ModeInfo comes from torchfunction or torchdispatch

Reply:
I think the stronger argument for having them merged is for the nested version (`push_{torch_function|python}_mode`). Because we're currently doing the mode determination in python, it would be great for any performance improvements or moves to C++ to benefit both types of modes. It seems likely that if I were writing an update, I would forget to update both of them

Having the merged implementations for this was kind of a product of merging the nested modes and then not wanting to have the unevenness of one of them being merged and one of them not. But open if it still feels like overkill

Updated ModeInfo though!

Reply:
I'm OK with linking up the designs here.

Reply:
+1, the deduplication sounds good and we can always un-deduplicate if it comes to that in the future